### ISSUE TITLE
CA1859 improperly handles explicit interface implementations on a parent class

### ISSUE THREAD

**Issue by jnyrup** [dotnet/roslyn-analyzers#6704]
### Analyzer

**Diagnostic ID**: [CA1859](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1859): `Use concrete types when possible for improved performance`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 8 Preview 5 SDK](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [8.0.100-preview.5.23303.2](https://dotnet.microsoft.com/download/dotnet/8.0)

### Describe the bug

#6416 described a false positive when a type explicitly implemented the interface method being invoked.
This issue is about when a _parent class_ explicitly implements the interface method.

### Steps To Reproduce

```csproj
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net7.0</TargetFramework>
    <AnalysisLevel>latest</AnalysisLevel>
  </PropertyGroup>
</Project>
```

```cs
class Class1
{
    public void NoWarning()
    {
        I obj = new Base();
        obj.M();
    }

    public void FalseWarning()
    {
        I obj = new Derived(); // <-- triggers CA1859
        obj.M();
    }
}

interface I
{
    void M();
}

class Base : I
{
    void I.M()
    {
    }
}

class Derived : Base
{
}

```

### Expected behavior
CA1859 should _not_ be triggered as `M()` is only explicitly implemented.

### Actual behavior
CA1859 is triggered.
```
CA1859	Change type of variable 'obj' from 'I' to 'Derived' for improved performance
```
